pkg/container-hook/tracer.go

```diff
@@ -0,0 +1,591 @@
+func initFanotify() (*fanotify.NotifyFD, error) {
+	fanotifyFlags := uint(unix.FAN_CLOEXEC | unix.FAN_CLASS_CONTENT | unix.FAN_UNLIMITED_QUEUE | unix.FAN_UNLIMITED_MARKS | unix.FAN_NONBLOCK)
+	openFlags := os.O_RDONLY | unix.O_LARGEFILE | unix.O_CLOEXEC
+	return fanotify.Initialize(fanotifyFlags, openFlags)
+}
+
+// NewContainerNotifier uses fanotify and ebpf to detect when a container is
+// created or terminated, and call the callback on such event.
+//
+// Limitations:
+// - the container runtime must be installed in one of the paths listed by runtimePaths
+func NewContainerNotifier(callback ContainerNotifyFunc) (*ContainerNotifier, error) {
+	n := &ContainerNotifier{
+		callback: callback,
+		done:     make(chan bool),
+	}
+
+	if err := n.install(); err != nil {
+		n.Close()
+		return nil, err
+	}
+
+	return n, nil
+}
+
+func (n *ContainerNotifier) install() error {
+	// Start fanotify
+	runtimeBinaryNotify, err := initFanotify()
+	if err != nil {
+		return err
+	}
+	n.runtimeBinaryNotify = runtimeBinaryNotify
+
+	// Load ebpf program
+	spec, err := loadExecruntime()
+	if err != nil {
+		return fmt.Errorf("load ebpf program for container-hook: %w", err)
+	}
+	if err := kallsyms.SpecUpdateAddresses(spec, []string{"fanotify_fops"}); err != nil {
+		return fmt.Errorf("RewriteConstants: %w", err)
+	}
+
+	gadgets.FixBpfKtimeGetBootNs(spec.Programs)
+	consts := map[string]interface{}{
+		"tracer_fanotify_fd": uint64(runtimeBinaryNotify.Fd),
+	}
+	if err := spec.RewriteConstants(consts); err != nil {
+		return fmt.Errorf("RewriteConstants: %w", err)
+	}
+
+	opts := ebpf.CollectionOptions{}
+
+	if err := spec.LoadAndAssign(&n.objs, &opts); err != nil {
+		return fmt.Errorf("loading maps and programs: %w", err)
+	}
+
+	// Initialize the ebpf programs by calling the iterator.
+	initIter, err := link.AttachIter(link.IterOptions{
+		Program: n.objs.IgFaIt,
+	})
+	if err != nil {
+		return fmt.Errorf("attach BPF iterator: %w", err)
+	}
+	defer initIter.Close()
+
+	initIterFile, err := initIter.Open()
+	if err != nil {
+		return fmt.Errorf("open BPF iterator: %w", err)
+	}
+	defer initIterFile.Close()
+	_, err = io.ReadAll(initIterFile)
+	if err != nil {
+		return fmt.Errorf("read BPF iterator: %w", err)
+	}
+
+	// Attach ebpf programs
+	l, err := link.Kprobe("fsnotify_remove_first_event", n.objs.IgFaPick, nil)
+	if err != nil {
+		return fmt.Errorf("attaching kprobe fsnotify_remove_first_event: %w", err)
+	}
+	n.links = append(n.links, l)
+
+	l, err = link.Kretprobe("fsnotify_remove_first_event", n.objs.IgFaPickX, nil)
+	if err != nil {
+		return fmt.Errorf("attaching kretprobe fsnotify_remove_first_event: %w", err)
+	}
+	n.links = append(n.links, l)
+
+	l, err = link.Tracepoint("syscalls", "sys_enter_execve", n.objs.IgExecveE, nil)
+	if err != nil {
+		return fmt.Errorf("attaching tracepoint: %w", err)
+	}
+	n.links = append(n.links, l)
+
+	l, err = link.Tracepoint("syscalls", "sys_exit_execve", n.objs.IgExecveX, nil)
```

Comment:
I think it needs to be different for arm64: Something like `link.Kretprobe("do_execveat_common.isra.0", objs.IgExecveatX, nil)`.